js/src/Axis.js

```diff
@@ -60,68 +53,78 @@ var Axis = widgets.WidgetView.extend({
             // TODO: rescale_axis does too many things. Decompose
             this.axis.scale(this.axis_scale.scale); // TODO: this is in redraw_axisline
             this.rescale_axis();
         }, this);
 
-        this.listenTo(this.model, "change:tick_values", this.set_tick_values, this);
-        this.listenTo(this.model, "change:tick_format", this.tickformat_changed, this);
-        this.listenTo(this.model, "change:tick_style", this.tickstyle_changed, this);
-        this.listenTo(this.model, "change:num_ticks", function(model, value) {
-            this.num_ticks = value;
-            this.set_tick_values();
-        }, this);
-        this.listenTo(this.model, "change:color", this.update_color, this);
-        this.model.on_some_change(["label", "label_color"], this.update_label, this);
-        this.model.on_some_change(["grid_color", "grid_lines"], this.update_grid_lines, this);
-        this.listenTo(this.model, "change:label_location", this.update_label_location, this);
-        this.listenTo(this.model, "change:label_offset", this.update_label_offset, this);
+        // Tick attributes
+        this.listenTo(this.model, "change:tick_values", this.update_label_settings, this);
+        this.listenTo(this.model, "change:tick_format", this.update_tick_format, this);
+        this.listenTo(this.model, "change:num_ticks", this.update_label_settings, this);
+        this.listenTo(this.model, "change:tick_transform", this.apply_tick_styling, this);
+        this.listenTo(this.model, "change:tick_style", this.apply_tick_styling, this);
+
+        // Label attributes
+        var labels = ["label_location", "label_offset", "label", "label_color"]
+        this.model.on_some_change(labels, this.update_label_settings, this)
+
+        // Axis attributes
+        this.listenTo(this.model, "change:color", this.update_line_color, this);
         this.listenTo(this.model, "change:visible", this.update_visibility, this);
+        this.listenTo(this.model, "change:offset", this.update_offset, this);
+        this.model.on_some_change(["grid_color", "grid_lines"], this.update_grid_lines, this);
         this.model.on_some_change(["side", "orientation"], this.update_display, this);
-        this.listenTo(this.model, "change:offset", function() {
-            var offset_creation_promise = this.get_offset();
-            var that = this;
-            offset_creation_promise.then(function() {
-                that.set_scales_range();
-                that.update_offset_scale_domain();
-                that.g_axisline.attr("transform", that.get_axis_transform());
-                that.update_grid_lines();
-            });
-        }, this);
         this.parent.on("margin_updated", this.parent_margin_updated, this);
     },
 
+    update_offset: function() {
+        var offset_creation_promise = this.get_offset_promise();
+        var that = this;
+        offset_creation_promise.then(function() {
```

Comment:
Should we use arrow functions? only IE11 doesn't support it?
cc @SylvainCorlay @jasongrout
Or should we keep this before moving to ES6+babel?

Reply:
> Or should we keep this before moving to ES6+babel?

Or Typescript, which also compiles down to either es5 or es6.

Reply:
I think moving to typescript is likely to make us lose contributors.

Maybe the separated scales are a good place for typescript.

Reply:
Yes, I'd say ES6 for code, typescript for unittests maybe?
But should we move to fat arrows functions yet?

Reply:
We can move to typescript as *just* as es6->es5 translator. Or we can just use es6 and insist the user has a recent browser. I'm just saying we don't have to use Babel. The nice thing about using TS as an es6->es5 translator is that it's easy to gradually introduce typing as it makes sense, and we also automatically get type checking for widgets stuff.

My guess is that just using TS to compile even our existing js will show us many bugs :). That's been my experience with converting codebases, anyway.

Reply:
That is true, it caught some bugs in ipyvolume as well, so stay with js, but use ts for compiling to es5 sounds like a good idea.